pkg/datapath/linux/node_ids.go

```diff
@@ -0,0 +1,152 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package linux
+
+import (
+	"net"
+
+	"github.com/sirupsen/logrus"
+
+	"github.com/cilium/cilium/api/v1/models"
+	"github.com/cilium/cilium/pkg/idpool"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+	"github.com/cilium/cilium/pkg/node"
+	nodeTypes "github.com/cilium/cilium/pkg/node/types"
+)
+
+const (
+	minNodeID = 1
+	maxNodeID = idpool.ID(^uint16(0))
+)
+
+// AllocateNodeID allocates a new ID for the given node (by IP) if one wasn't
+// already assigned.
+func (n *linuxNodeHandler) AllocateNodeID(nodeIP net.IP) uint16 {
+	n.mutex.Lock()
+	defer n.mutex.Unlock()
+	return n.allocateNodeIDLocked(nodeIP)
```

Comment:
I think we often have this because the `funcLocked` version needs to be called elsewhere directly. See for example `deallocateNodeIDLocked` below.

That however doesn't appear to be the case here. I'll change to remove `allocateNodeIDLocked`. I probably need it in a previous version of the code and didn't revert then.